Seo for Messaging (#671)
apache/pulsar-site: docs/concepts-messaging.md

### Batching

> In Pulsar, batching means the producer accumulates and sends a batch of messages in a single request. The batch size is defined by the maximum number of messages and the maximum publish latency. Therefore, the backlog size represents the total number of batches instead of the total number of messages.

Reviewer: Suggest keeping the original text unchanged since it's accurate

> ![Batching mechanism in Pulsar](/assets/batching.svg)

Reviewer: ```suggestion
![Batching in Pulsar](/assets/batching.svg)
```

![How Broker stores chunked message when there is only one producer](/assets/chunking-01.png)

#### Handle interwoven chunked messages with one ordered consumer

When multiple producers publish chunked messages into a single topic, the broker stores all the chunked messages coming from different producers in the same [managed ledger](concepts-architecture-overview.md#managed-ledgers). The chunked messages in the managed ledger can be interwoven with each other. As shown below, Producer 1 publishes message M1 in three chunks M1-C1, M1-C2 and M1-C3. Producer 2 publishes message M2 in three chunks M2-C1, M2-C2 and M2-C3. All chunked messages of the specific message are still in order but might not be consecutive in the managed ledger.

> ![How Broker stores chunked message when there are mutiple producers](/assets/chunking-02.png)

Reviewer: ```suggestion
![Interwoven chunked messages in Pulsar](/assets/chunking-02.png)
```

## Namespaces

Pulsar namespaces are logical groupings of topics. A namespace is a logical nomenclature within a tenant. A tenant creates namespaces via the [admin API](admin-api-namespaces.md#create-namespaces). For instance, a tenant with different applications can create a separate namespace for each application. A namespace allows the application to create and manage a hierarchy of topics. The topic `my-tenant/app1` is a namespace for the application `app1` for `my-tenant`. You can create any number of [topics](#topics) under the namespace.

## Subscriptions

> A Pulsar subscription is a lease on a topic established by a group of consumers, it is a named configuration rule that determines how messages are delivered to consumers. There are four available subscription types in Pulsar:

Reviewer: same as https://github.com/apache/pulsar-site/pull/671#discussion_r1287836845

> A Pulsar subscription is a lease on a topic established by a group of consumers, it is a named configuration rule that determines how messages are delivered to consumers. There are four available subscription types in Pulsar:

Reviewer: ```suggestion
A Pulsar subscription is a named configuration rule that determines how messages are delivered to consumers. It is a lease on a topic established by a group of consumers. There are four subscription types in Pulsar:
```

- [exclusive](#exclusive)
- [shared](#shared)
- [failover](#failover)
- [key_shared](#key_shared) 

These types are illustrated in the figure below.

> ![Four Subscription types in Pulsar](/assets/pulsar-subscription-types.png)

Reviewer: ```suggestion
![Subscription types in Pulsar](/assets/pulsar-subscription-types.png)
```

#### Exclusive

> The Exclusive type is a subscription mode that only allows a single consumer to attach to the subscription. If multiple consumers subscribe to a topic using the same subscription, an error occurs. Note that if the topic is partitioned, all partitions will be consumed by the single consumer allowed to be connected to the subscription.

Reviewer: ```suggestion
The exclusive type is a subscription type that only allows a single consumer to attach to the subscription. If multiple consumers subscribe to a topic using the same subscription, an error occurs. Note that if the topic is partitioned, all partitions will be consumed by the single consumer allowed to be connected to the subscription.
```

:::

> ![Pulsar Exclusive subscriptions](/assets/pulsar-exclusive-subscriptions.svg)

Reviewer: ```suggestion
![Exclusive subscription type in Pulsar](/assets/pulsar-exclusive-subscriptions.svg)
```

#### Failover

> The Failover type is a subscription mode that multiple consumers can attach to the same subscription. 

Reviewer: ```suggestion
The failover type is a subscription type that multiple consumers can attach to the same subscription. 
```

      
      - for P1: Consumer D is the active consumer and Consumer C is the stand-by consumer.

>   ![Workflow of Pulsar Failover subscriptions when the number of partitions in a partitioned topic is less than the number of consumers](/assets/pulsar-failover-subscriptions-5.png)

Reviewer: ```suggestion
  ![Workflow of failover subscription type in Pulsar](/assets/pulsar-failover-subscriptions-5.png)
```

  
  - P2, P5, and P8 are assigned to Consumer C. Consumer C is their active consumer. Consumer A and Consumer B are their stand-by consumers.
  

>   ![Workflow of Pulsar Failover subscriptions when the number of partitions in a partitioned topic is greater than the number of consumers](/assets/pulsar-failover-subscriptions-1.svg)

Reviewer: ```suggestion
  ![Workflow of failover subscription type in Pulsar](/assets/pulsar-failover-subscriptions-1.svg)
```
Keep it as concise as possible

  
  Consumer A is the master consumer, while consumer B would be the next consumer in line to receive messages if consumer A is disconnected.

>   ![Workflow of Pulsar Failover subscriptions when there is one non-partitioned topic](/assets/pulsar-failover-subscriptions-2.svg)

Reviewer: ```suggestion
  ![Workflow of failover subscription type in Pulsar](/assets/pulsar-failover-subscriptions-2.svg)
```

  
  - The non-partitioned topic 2 and non-partitioned topic 3 are assigned to consumer A. Consumer B is their stand-by consumer.

>   ![Workflow of Pulsar Failover subscriptions when there are multiple non-partitioned topics](/assets/pulsar-failover-subscriptions-3.svg)

Reviewer: ```suggestion
  ![Workflow of failover subscription type in Pulsar](/assets/pulsar-failover-subscriptions-3.svg)
```

#### Shared

> The Pulsar Shared subscriptions allow multiple consumers attach to the same subscription. Messages are delivered in a round-robin distribution across consumers, and any given message is delivered to only one consumer. When a consumer disconnects, all the messages that were sent to it and not acknowledged will be rescheduled for sending to the remaining consumers.

Reviewer: ```suggestion
The shared subscription type in Pulsar allows multiple consumers to attach to the same subscription. Messages are delivered in a round-robin distribution across consumers, and any given message is delivered to only one consumer. When a consumer disconnects, all the messages that were sent to it and not acknowledged will be rescheduled for sending to the remaining consumers.
```

:::

> ![Pulsar Shared subscriptions](/assets/pulsar-shared-subscriptions.svg)

Reviewer: ```suggestion
![Shared subscription type in Pulsar](/assets/pulsar-shared-subscriptions.svg)
```

#### Key_Shared

> The Pulsar Key_Shared subscriptions allow multiple consumers attach to the same subscription. But different with the Shared type, messages in Key_Shared type are delivered in distribution across consumers and messages with the same key or same ordering key are delivered to only one consumer. No matter how many times the message is re-delivered, it is delivered to the same consumer.

Reviewer: ```suggestion
The Key_Shared subscription type in Pulsar allows multiple consumers to attach to the same subscription. But different with the Shared type, messages in the Key_Shared type are delivered in distribution across consumers and messages with the same key or same ordering key are delivered to only one consumer. No matter how many times the message is re-delivered, it is delivered to the same consumer.
```

> ![Pulsar Key_Shared subscriptions](/assets/pulsar-key-shared-subscriptions.svg)

Reviewer: ```suggestion
![Key_Shared subscription type in Pulsar](/assets/pulsar-key-shared-subscriptions.svg)
```

## Partitioned topics

Normal topics are served only by a single broker, which limits the maximum throughput of the topic. Partitioned topic is a special type of topic handled by multiple brokers, thus allowing for higher throughput.

A partitioned topic is implemented as N internal topics, where N is the number of partitions. When publishing messages to a partitioned topic, each message is routed to one of several brokers. The distribution of partitions across brokers is handled automatically by Pulsar.

The diagram below illustrates this:

> ![How Pulsar handles the distribution of partitions across brokers](/assets/partitioning.png)

Reviewer: ```suggestion
![Partition distribution in Pulsar](/assets/partitioning.png)
```

### Performance

> With persistent topics, all messages are durably persisted on disks, whereas with non-persistent topics, brokers don't persist messages and immediately send acks back to the producer as soon as that message is delivered to connected brokers, so Non-persistent messaging is usually faster than persistent messaging. Producers thus see comparatively low publish latency with non-persistent topic.

Reviewer: ```suggestion
With persistent topics, all messages are durably persisted on disks, whereas with non-persistent topics, brokers don't persist messages and immediately send acks back to the producer as soon as that message is delivered to connected brokers, so non-persistent messaging is usually faster than persistent messaging. Producers thus see comparatively low publish latency with non-persistent topics.
```

- Consumers get disconnected from the database or the HTTP server. When this happens, the database is temporarily offline while the consumer is writing the data to it and the external HTTP server that the consumer calls are momentarily unavailable.
- Consumers get disconnected from a broker due to consumer crashes, broken connections, etc. As a consequence, unacknowledged messages are delivered to other available consumers.

> Message redelivery in Apache Pulsar avoids these and other message delivery failures using at-least-once delivery semantics that ensure Pulsar processes a message more than once.

Reviewer: What does "these" refer to? Please specify them explicitly

### Producer idempotency

> The other available approach is typically called **producer idempotency**, that is, each message is *only produced once* without data loss and duplicatuion. The drawback of this approach is that it defers the work of message deduplication to the application. In Pulsar, this is handled at the [broker](reference-terminology.md#broker) level, so you do not need to modify your Pulsar client code. Instead, you only need to make administrative changes. For details, see [Managing message deduplication](cookbooks-deduplication.md).

Reviewer: ```suggestion
The other available approach to message deduplication is **producer idempotency**, which means each message is *only produced once* without data loss and duplication. The drawback of this approach is that it defers the work of message deduplication to the application. In Pulsar, this is handled at the [broker](reference-terminology.md#broker) level, so you do not need to modify your Pulsar client code. Instead, you only need to make administrative changes. For details, see [Managing message deduplication](cookbooks-deduplication.md).
```
